pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java

```diff
@@ -129,6 +130,11 @@ private void assignWorkersToLeafFragment(PlanFragment fragment, DispatchablePlan
         Preconditions.checkState(tableTypeToSegmentListMap.put(tableType, serverEntry.getValue()) == null,
             "Entry for server {} and table type: {} already exist!", serverEntry.getKey(), tableType);
       }
+
+      // attach unavailable segments to metadata
+      if (!routingTable.getUnavailableSegments().isEmpty()) {
+        metadata.addTableToUnavailableSegmentsMap(logicalTableName, routingTable.getUnavailableSegments());
+      }
```

Comment:
cant we simply attach to `DispatchablePlanContext` this way no need to change anything in fragment

Reply:
In the last step of planner, we need to construct DispatchableSubPlan from PlanFragment subPlanRoot and  DispatchablePlanContext dispatchablePlanContext
```
private static DispatchableSubPlan finalizeDispatchableSubPlan(PlanFragment subPlanRoot,
      DispatchablePlanContext dispatchablePlanContext) {
    return new DispatchableSubPlan(dispatchablePlanContext.getResultFields(),
        dispatchablePlanContext.constructDispatchablePlanFragmentList(subPlanRoot),
        dispatchablePlanContext.getTableNames());
  }
```

So if we don't do it here, then we need to add a new field into `DispatchableSubPlan`

Reply:
yeah adding a new field to dispatchable subplan is what i meant. 

e.g. 
1. `WorkerManager` here attach all unavailable segments to `dispatchablePlanContext._unavailableSegmentMap`
2. `PinotDispatchPlanner#finalizeDispatchableSubPlan` copies the map to `DispachableSubPlan._unavailableSegmentMap`

there's no need for planFragment to know this as they cannot do anything anyway. right?

Reply:
make sense.